Handle file upload from post request

AnonymousWalker opened the pull request:
I used a map to store data for each "PartData.FormItem" as parameters. Looking for a better approach.
I also replace the FilePathTestModel with FileUploadModel so that I can pass it directly to the algorithm function "createPathFromFile(...)"

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/anonymouswalker/safileupload/2)
<!-- Reviewable:end -->

16dprice committed:
update js and html to post a form with vanilla js

AnonymousWalker committed:
handle file upload from server

AnonymousWalker committed:
remove unused files

AnonymousWalker committed:
Refactor post request file handler

16dprice commented on src/js/main.js:
```diff
@@ -1,33 +1,19 @@
-        .then(res => handleResponse(res))
-        .catch(err => console.log(err));
-}
+        .then(res => console.log(res)) // TODO: handle the response appropriately
```
we'll have to handle this before approving the pull request, of course

16dprice commented on src/js/main.js:
```diff
@@ -1,33 +1,19 @@
-        $(".output-container").removeClass("error")
-        $(".output-container").addClass("success")
+        $(".main-grid__output").removeClass("main-grid__output--error").addClass("main-grid__output--success")
     } else {
-        $("#output").text(res.data.output)
-        $(".output-container").removeClass("success")
-        $(".output-container").addClass("error")
+        $(".main-grid__output").removeClass("main-grid__output--success").addClass("main-grid__output--error")
```
Actually, these classes are invalid because they are in the CSS file in another pull request. This should be fine though once the other pull request is approved and we can merge dev into this branch before approving the pull request.

16dprice commented on src/main/kotlin/AppModule.kt:
```diff
@@ -47,49 +50,65 @@ fun Application.module() {
         route("/") {
             get {
-                call.respond(ThymeleafContent("index", mapOf("obj" to "")))
+                call.respond(ThymeleafContent("index", mapOf("obj" to ""), locale = Locale("vi")))
```
this shouldn't default to Vietnamese

AnonymousWalker replied:
Oh yes I've got used to that for a while. Now changing back

16dprice commented on src/main/kotlin/AppModule.kt:
```diff
@@ -47,49 +50,65 @@ fun Application.module() {
+            val multiPart = call.receiveMultipart()
+            val dataMap: MutableMap<String, String> = mutableMapOf()
+            val fileUploadModel: FileUploadModel
+            var file = File("")
```
maybe it's better to not declare this here? and just declare file as a val variable on line 68 instead.

AnonymousWalker replied:
I tried that but that "file" variable should be out of the scope of multiPart.forEachPart block.

16dprice replied:
maybe that will change if you move things out into a different class. You could store the file as a class property and access it that way? It would probably make things cleaner overall

16dprice commented on src/main/kotlin/AppModule.kt:
```diff
@@ -47,49 +50,65 @@ fun Application.module() {
+    val model = if (data["mediaQuality"]!!.isNotBlank()) {
+        FileUploadModel(
+            inputFile = file,
+            languageCode = data["languageCode"]!!,
+            dublinCoreId = data["dublinCoreId"]!!,
+            grouping = data["grouping"]!!,
+            projectId = data["projectId"]!!,
+            mediaExtension = data["mediaExtension"]!!,
+            mediaQuality = data["mediaQuality"]!!
```
I don't think you need the double exclamation points because our validation will check for correct values anyway

AnonymousWalker replied:
That is true. However, when accessing a key-values in this map, it would return a nullable String? in this case, so I had to put "!!" there. I may need to find a better approach to store these params...

16dprice commented on src/main/kotlin/AppModule.kt:
```diff
@@ -47,49 +50,65 @@ fun Application.module() {
-            mediaExtension = data.mediaExtension,
-            mediaQuality = data.mediaQuality,
-            expectedResult = ""
+private fun processFileUpload(file: File, data: Map<String,String>): String {
```
I think that this function may need to be in another file. It seems clunky to have the functionality to handle the process in the same place that the route for the post request is being handled. I would consider only receiving the data with the call.receiveMultipart() function and then taking that and feeding it into another class. That class could then have methods to process the file upload and give a response.

AnonymousWalker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @16dprice and @AnonymousWalker)

___
*[src/main/kotlin/AppModule.kt, line 86 at r2](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2MiN9uR-QQqxlHtdH-r2-86:-MA2Q7sS-lmWxju_mUYh:b-vvk8dg) ([raw file](https://github.com/anonymouswalker/safileupload/blob/4f33dd4d97c0b6d959c95ce48a37837027b0e015/src/main/kotlin/AppModule.kt#L86)):*
<details><summary><i>Previously, 16dprice wrote…</i></summary><blockquote>

I think that this function may need to be in another file. It seems clunky to have the functionality to handle the process in the same place that the route for the post request is being handled. I would consider only receiving the data with the call.receiveMultipart() function and then taking that and feeding it into another class. That class could then have methods to process the file upload and give a response.
</blockquote></details>

Sure! I agree


<!-- Sent from Reviewable.io -->

AnonymousWalker commented:
___
*[src/js/main.js, line 9 at r2](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2MiN9uR-QQqxlHtdD-r2-9:-MA2b6TF5caKPDDuIuXv:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/4f33dd4d97c0b6d959c95ce48a37837027b0e015/src/js/main.js#L9)):*
<details><summary><i>Previously, 16dprice wrote…</i></summary><blockquote>

we'll have to handle this before approving the pull request, of course
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AnonymousWalker committed:
refactor /upload post handler into RequestHandler.kt

16dprice commented on src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt:
```diff
@@ -0,0 +1,76 @@
+
+    fun handleFormUpload(parts: List<PartData>): String {
+        val paramsMap: MutableMap<String, String> = mutableMapOf()
+        var file = File("./src/main/resources/uploads/")
```
I still think this should probably be a property of the class because eventually you will need to save the file and I think it will be best if that file is available to the whole class.

AnonymousWalker replied:
Do you mean a data class with only one file property in it?

16dprice replied:
No, I think that the file should be in this class, but it could be declared as a private variable i.e. it would be private val file: File
It would be initialized once the handleFormUpload function is called. Or you could perhaps give the class a constructor that accepts a file as input? That may be over complicating things though

16dprice commented on src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private fun processUploadModel(file: File, params: Map<String, String>): String {
+        var response = mutableMapOf<String, Any>()
```
I believe you can change this to be val instead of var. This is because when you edit the variable, you are actually changing the map object that the variable points to in memory. You don't actually edit the variable in any way (i.e. you don't change where it is pointing so response never actually changes, technically).

AnonymousWalker replied:
Yeah I missed that one when working with mutablemap and stuff

16dprice commented on src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt:
```diff
@@ -0,0 +1,76 @@
+                    mediaExtension = params["mediaExtension"]!!
+                )
+            }
+            response.putAll(
+                mapOf(
+                    "output" to FilePathGenerator.createPathFromFile(model),
+                    "success" to true
+                )
+            )
+        } catch (ex: IllegalArgumentException) {
+            response.putAll(
+                mapOf(
+                    "output" to ex.message!!,
+                    "success" to false
+                )
+            )
```
I would suggest having two different functions that handle getting the response, or perhaps just the map that is being put into the response. I suggest this because that function can then check to see if it will succeed and then save the file that I think should be a class property before returning that the process succeeded.

AnonymousWalker replied:
Yeah I missed that one when working with mutablemap and stuff

16dprice commented on src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt:
```diff
@@ -0,0 +1,76 @@
+
+    private fun processUploadModel(file: File, params: Map<String, String>): String {
+        var response = mutableMapOf<String, Any>()
+        try {
```
You have a large try body because the FileUploadModel validates upon initialization. Maybe we should change this? It's quite inconvenient to have to catch errors upon instantiating a data class. Let me know your thoughts on that. If we didn't validate upon instantiation, it would also make it easier to break up the rest of this function into more, smaller functions.

16dprice commented on src/main/kotlin/AppModule.kt:
```diff
@@ -50,75 +42,15 @@ fun Application.module() {
-                call.respond(ThymeleafContent("index", mapOf("obj" to ""), locale = Locale("vi")))
+                call.respond(ThymeleafContent("index", mapOf("obj" to "")))
             }
         }
```
I think it's possible (and maybe preferred) to change this to
get("/") {
...
}
instead of
route("/") {
    get {
        ....
    }
}

AnonymousWalker committed:
refactor RequestHandler

AnonymousWalker committed:
replace route() with get()

AnonymousWalker committed:
small refactor

AnonymousWalker committed:
rename from "dublinCoreId" to "resourceType"

AnonymousWalker committed:
remove uncessary files & redundant code

16dprice committed:
using selects where applicable instead of inputs

AnonymousWalker committed:
Remove gson. Uploaded file will be copied to a correct directory. Refactor exception handler handler

Handle both IllegalArgumentException & FileSystemException (file already exists)

AnonymousWalker committed:
Merge branch 'tony-file_upload' of https://github.com/AnonymousWalker/SAFileUpload into tony-file_upload

AnonymousWalker commented on src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt:
```diff
@@ -35,23 +37,30 @@ class RequestHandler {
+
+            output = FilePathGenerator.createPathFromFile(model)
+
+            val newDirectory = File("./src/main/resources/uploads/$output")
```
./src/main/resources/uploads/ 
This will be replaced with server's root directory. We'll try to put it in a key-value in a config file

jsarabia reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 13 files reviewed, 17 unresolved discussions (waiting on @16dprice, @AnonymousWalker, and @jsarabia)

___
*[src/js/main.js, line 9 at r2](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2MiN9uR-QQqxlHtdD-r2-9:-MA7zGb03M6tfZVaFEl3:bk7xs85) ([raw file](https://github.com/anonymouswalker/safileupload/blob/4f33dd4d97c0b6d959c95ce48a37837027b0e015/src/js/main.js#L9)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

formatting looks like the . of the .then should be aligned with the } on the line above it
___
*[src/main/kotlin/Startup.kt, line 8 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA7wVMW7USJuT1qjTlT:-MA7wVMXBFG4pvic7sAw:b-m4hp1f) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/Startup.kt#L8)):*
> ```Kotlin
>     embeddedServer(
>         Netty,
>         4567,
> ```

can we declare this as a constant?
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 15 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA80tf4EwlBE1bpLuth:-MA80tf4EwlBE1bpLuti:bpfrnbx) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L15)):*
> ```Kotlin
> 
>     fun handleFormUpload(parts: List<PartData>): String {
>         val paramsMap: MutableMap<String, String> = mutableMapOf()
> ```

using filter and or associateBy to construct this map may make this more readable.
Maybe make two lists from the parts:
fileItems = parts.filter { it is FileItem }
formItems = parts.filter { it is FormItem }

something like that?
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 69 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-THGBLULLOW0MduO:-MA8-THH5CRP4Ka1n7YB:bnr84dq) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L69)):*
> ```Kotlin
>     @Throws(IllegalArgumentException::class)
>     private fun createModel(file: File, params: Map<String, String>): FileUploadModel {
>         return if (params["mediaQuality"]!!.isNotBlank()) {
> ```

These if/else statements are massive, because the object has a lot of parameters. Can we try to make an if/else maybe be just one line of one of these parameters? Or can we refactor the Model object in some way such that construction of it is less clunky and verbose? Maybe it needs a builder?
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 77 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-y7g-anlxc7cwzV0:-MA8-y7g-anlxc7cwzV1:b-qb5h0v) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L77)):*
> ```Kotlin
>                 projectId = params["projectId"]!!,
>                 mediaExtension = params["mediaExtension"]!!,
>                 mediaQuality = params["mediaQuality"]!!
> ```

!! means these aren't null, but I'm not seeing any validation to prove these keys exist, so we're allowing for crashes here.
___
*[src/main/resources/templates/index.html, line 36 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA7zqro4E3NDC-DB_Xl:-MA7zqro4E3NDC-DB_Xm:bbsybqr) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/resources/templates/index.html#L36)):*
> ```HTML
>                 <option value="">N/A</option>
>                 <option value="mp3">mp3</option>
>                 <option value="wav">wav</option>
> ```

Let's move the actual list into our kotlin code and have the template iterate over the list to generate the values here. (applies to all lists like this)
___
*[src/main/resources/templates/index.html, line 48 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-7vS7d0h2DfnODcb:-MA8-7vS7d0h2DfnODcc:b-res1ji) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/resources/templates/index.html#L48)):*
> ```HTML
>             <label for="grouping"><span th:text="#{grouping}">Grouping</span><sup>*</sup></label>
>             <select name="grouping" id="grouping">
>                 <option value="book">Book</option>
> ```

How would you localize "Book"?


<!-- Sent from Reviewable.io -->

AnonymousWalker commented:
___
*[src/main/kotlin/Startup.kt, line 8 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA7wVMW7USJuT1qjTlT:-MA87s918LE5W8wOLj6P:bxqxsm) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/Startup.kt#L8)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

can we declare this as a constant?
</blockquote></details>

Yes, we're trying to find a way to put these constants into a config file. (application.conf looks good but I have no idea how to include it into the server config)


<!-- Sent from Reviewable.io -->

AnonymousWalker commented:
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 15 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA80tf4EwlBE1bpLuth:-MA8Dd7i4pDqrCb6ziMv:b-yttrka) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L15)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

using filter and or associateBy to construct this map may make this more readable.
Maybe make two lists from the parts:
fileItems = parts.filter { it is FileItem }
formItems = parts.filter { it is FormItem }

something like that?
</blockquote></details>

nice! I was looking for this way of getting the map. The associateBy is so helpful 


<!-- Sent from Reviewable.io -->

AnonymousWalker commented:
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 77 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-y7g-anlxc7cwzV0:-MA8Ee6u4rK4fjRPtTzK:b-ycouop) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L77)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

!! means these aren't null, but I'm not seeing any validation to prove these keys exist, so we're allowing for crashes here.
</blockquote></details>

yeah I'll update it with  ?: ""


<!-- Sent from Reviewable.io -->

AnonymousWalker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 13 files reviewed, 17 unresolved discussions (waiting on @16dprice, @AnonymousWalker, and @jsarabia)

___
*[src/js/main.js, line 17 at r2](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2MiN9uR-QQqxlHtdD-r2-17:-MA2b5qD5y1j5yi1kPu0:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/4f33dd4d97c0b6d959c95ce48a37837027b0e015/src/js/main.js#L17)):*
<details><summary><i>Previously, 16dprice wrote…</i></summary><blockquote>

Actually, these classes are invalid because they are in the CSS file in another pull request. This should be fine though once the other pull request is approved and we can merge dev into this branch before approving the pull request.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AnonymousWalker committed:
Rewrite map construction with associateBy, filter from list of FormItems

AnonymousWalker committed:
refactor createModel (big if/else -> single line)

16dprice committed:
more internationalization

AnonymousWalker committed:
Change upload model that mediaQuality can accept empty string

If mediaQuality empty, set it to "hi" at _init_

AnonymousWalker committed:
Merge branch 'tony-file_upload' of https://github.com/AnonymousWalker/SAFileUpload into tony-file_upload

16dprice committed:
get preferred language from accept-language http header

AnonymousWalker committed:
add missing keyword in Enum classes

AnonymousWalker committed:
Create template list for <select> options from server. Rewrite localization properties files

AnonymousWalker committed:
move js/css files to resources package, embeded to application and serve index.html reference.

replace "localhost:.../static/..."
with "static/..."

AnonymousWalker committed:
move files to packages, remove unused import

16dprice committed:
getting the preferred locale using native kotlin and java methods

AnonymousWalker committed:
minor changes

AnonymousWalker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 19 files reviewed, 17 unresolved discussions (waiting on @16dprice, @AnonymousWalker, and @jsarabia)

___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 15 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA80tf4EwlBE1bpLuth:-MAgx3Qj5hDxKF5m5JYy:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L15)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

nice! I was looking for this way of getting the map. The associateBy is so helpful 
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 69 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-THGBLULLOW0MduO:-MAgx6qDA3guol44Emgf:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L69)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

These if/else statements are massive, because the object has a lot of parameters. Can we try to make an if/else maybe be just one line of one of these parameters? Or can we refactor the Model object in some way such that construction of it is less clunky and verbose? Maybe it needs a builder?
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 77 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-y7g-anlxc7cwzV0:-MAgx7fL7h2sliO_7iZE:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L77)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

yeah I'll update it with  ?: ""
</blockquote></details>

Done.
___
*[src/main/resources/templates/index.html, line 36 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA7zqro4E3NDC-DB_Xl:-MAgxDJ-ESxZI_idvuTm:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/resources/templates/index.html#L36)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

Let's move the actual list into our kotlin code and have the template iterate over the list to generate the values here. (applies to all lists like this)
</blockquote></details>

Done.
___
*[src/main/kotlin/Startup.kt, line 8 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA7wVMW7USJuT1qjTlT:-MAgx9ZX2zdIVc9BsgAD:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/kotlin/Startup.kt#L8)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

Yes, we're trying to find a way to put these constants into a config file. (application.conf looks good but I have no idea how to include it into the server config)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AnonymousWalker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 19 files reviewed, 17 unresolved discussions (waiting on @16dprice, @AnonymousWalker, and @jsarabia)

___
*[src/main/resources/templates/index.html, line 48 at r6](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA8-7vS7d0h2DfnODcb:-MAgx_YACYuYGfxHE8Ng:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/deceb5139e5b673e41b3d32fc3bec736fae452fa/src/main/resources/templates/index.html#L48)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

How would you localize "Book"?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AnonymousWalker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 19 files reviewed, 11 unresolved discussions (waiting on @16dprice and @jsarabia)

___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 14 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-14:-MAgy1wY-KpRuXmxqpc0:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L14)):*
<details><summary><i>Previously, 16dprice wrote…</i></summary><blockquote>

No, I think that the file should be in this class, but it could be declared as a private variable i.e. it would be private val file: File
It would be initialized once the handleFormUpload function is called. Or you could perhaps give the class a constructor that accepts a file as input? That may be over complicating things though
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 37 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-37:-MAgy2n12hT8hofmktsp:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L37)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

Yeah I missed that one when working with mutablemap and stuff
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 38 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-38:-MAgy3Nh2c45w67uH757:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L38)):*
<details><summary><i>Previously, 16dprice wrote…</i></summary><blockquote>

You have a large try body because the FileUploadModel validates upon initialization. Maybe we should change this? It's quite inconvenient to have to catch errors upon instantiating a data class. Let me know your thoughts on that. If we didn't validate upon instantiation, it would also make it easier to break up the rest of this function into more, smaller functions.
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 71 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-71:-MAgy4TkAYoj_an2LFUO:b-896fix) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L71)):*
<details><summary><i>Previously, 16dprice wrote…</i></summary><blockquote>

I would suggest having two different functions that handle getting the response, or perhaps just the map that is being put into the response. I suggest this because that function can then check to see if it will succeed and then save the file that I think should be a class property before returning that the process succeeded.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jsarabia reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 19 files reviewed, 11 unresolved discussions (waiting on @16dprice, @AnonymousWalker, and @jsarabia)

___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 14 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-14:-MAh2XYD1uFPmjocZZjz:b-bq0hog) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L14)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

this is going to look for a file at the root of the directory of where the process is launched.

As in, there's not going to be a src directory.

Access the resource from the getResource() method.
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/model/FileUploadModel.kt, line 21 at r14](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MAh4FYD5Tsp1NKEhOl9:-MAh4FYD5Tsp1NKEhOlA:b-9ku7kj) ([raw file](https://github.com/anonymouswalker/safileupload/blob/175b204e00c0b967336307788fe69adc5ad9d26a/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/model/FileUploadModel.kt#L21)):*
> ```Kotlin
> 
>     init {
>         if(this.mediaQuality.isEmpty()) this.mediaQuality = "hi"
> ```

feels like maybe this should go in validate
___
*[src/js/main.js, line 9 at r2](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2MiN9uR-QQqxlHtdD-r2-9:-MAh4OO5Dko_WTdRCYRS:b-x4jztm) ([raw file](https://github.com/anonymouswalker/safileupload/blob/4f33dd4d97c0b6d959c95ce48a37837027b0e015/src/js/main.js#L9)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

formatting looks like the . of the .then should be aligned with the } on the line above it
</blockquote></details>

@anonymouswalker ^
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/AppModule.kt, line 69 at r14](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MAh0Nxy1k6_752AavPH:-MAh0Ny-Aj_wACE0zjsY:b9th3ap) ([raw file](https://github.com/anonymouswalker/safileupload/blob/175b204e00c0b967336307788fe69adc5ad9d26a/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/AppModule.kt#L69)):*
> ```Kotlin
>         val locale = Locale.Builder().setLanguageTag(languageRange.range).build()
>         try {
>             ResourceBundle.getBundle("templates/index", locale, noFallbackController)
> ```

do you not want to try to test the rest of their preferences before just falling back to the default?

should "index" be a parameter to this function so that it can be reused for providing the locale of other templates?
___
*[src/main/resources/js/main.js, line 13 at r14](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MAh4Xmq0z8qNKRF9KZA:-MAh4Xmr7oWxasTO1kBk:b-wydmii) ([raw file](https://github.com/anonymouswalker/safileupload/blob/175b204e00c0b967336307788fe69adc5ad9d26a/src/main/resources/js/main.js#L13)):*
> ```JavaScript
> 
> function handleResponse(res) {
>     $(".main-grid__output-text").text(res.data.output)
> ```

Document.querySelector?
___
*[src/main/resources/js/main.js, line 15 at r14](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MAh4csW0oXYCuMeWctp:-MAh4csW0oXYCuMeWctq:bqwbscp) ([raw file](https://github.com/anonymouswalker/safileupload/blob/175b204e00c0b967336307788fe69adc5ad9d26a/src/main/resources/js/main.js#L15)):*
> ```JavaScript
>     $(".main-grid__output-text").text(res.data.output)
>     if(res.data.success) {
>         $(".main-grid__output").removeClass("main-grid__output--error").addClass("main-grid__output--success")
> ```

you'll need to move the . methods on their own lines for line length


<!-- Sent from Reviewable.io -->

16dprice commented:
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/AppModule.kt, line 69 at r14](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MAh0Nxy1k6_752AavPH:-MAh9fe66ITq7F1wUEkQ:b-ul29hv) ([raw file](https://github.com/anonymouswalker/safileupload/blob/175b204e00c0b967336307788fe69adc5ad9d26a/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/AppModule.kt#L69)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

do you not want to try to test the rest of their preferences before just falling back to the default?

should "index" be a parameter to this function so that it can be reused for providing the locale of other templates?
</blockquote></details>

This does test all of their preferences before falling back to the default. The variable languageRanges is an ordered list of the preferred language ranges and the function attempts to load a resource bundle based on that locale. If the the resource bundle loaded (i.e. didn't throw an error), then it will return that locale, which will be the one they've given the most priority. If none of the languages in languageRanges has a resource bundle, then it will return the default. Is this what you meant?

Yes, "index" should be a parameter to the function. That definitely makes sense.


<!-- Sent from Reviewable.io -->

AnonymousWalker committed:
Merge branch 'dev' into tony-file_upload

16dprice committed:
update js to handle response appropriately

16dprice committed:
remove axios dependency

16dprice committed:
add template name to function that gets the preferred locale

AnonymousWalker commented:
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 14 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-14:-MAhGcp8D3KPOIQwiaDZ:b4csb0r) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L14)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

this is going to look for a file at the root of the directory of where the process is launched.

As in, there's not going to be a src directory.

Access the resource from the getResource() method.
</blockquote></details>

I got it! Thanks.


<!-- Sent from Reviewable.io -->

AnonymousWalker commented:
___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/model/FileUploadModel.kt, line 21 at r14](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MAh4FYD5Tsp1NKEhOl9:-MAhGmDL-wO3glj00EGx:b-bqk2zn) ([raw file](https://github.com/anonymouswalker/safileupload/blob/175b204e00c0b967336307788fe69adc5ad9d26a/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/model/FileUploadModel.kt#L21)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

feels like maybe this should go in validate
</blockquote></details>

Will do!


<!-- Sent from Reviewable.io -->

AnonymousWalker committed:
use getResource() for accessing resources dir

javaClass.getResource(...)

AnonymousWalker committed:
Merge branch 'tony-file_upload' of https://github.com/AnonymousWalker/SAFileUpload into tony-file_upload

AnonymousWalker committed:
fix bug UI: output message block overlapped the submit button

CSS: display set to none
minor update: change timeout to 5s

jsarabia reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/anonymouswalker/safileupload/2)* status: 0 of 18 files reviewed, 8 unresolved discussions (waiting on @16dprice, @AnonymousWalker, and @jsarabia)

___
*[src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt, line 14 at r3](https://reviewable.io/reviews/anonymouswalker/safileupload/2#-MA2kM_gxZZozpHZAx_r-r3-14:-MAlcFQBBUvyPKDQKMii:brocunt) ([raw file](https://github.com/anonymouswalker/safileupload/blob/73a17a13416e98510a0236b653375903a7888b49/src/main/kotlin/org/wycliffeassociates/sourceaudio/upload/RequestHandler.kt#L14)):*
<details><summary><i>Previously, AnonymousWalker (Tony) wrote…</i></summary><blockquote>

I got it! Thanks.
</blockquote></details>

https://docs.oracle.com/javase/8/docs/technotes/guides/lang/resources.html

You're getting the resource as a URL and then taking the path of that. My expectation is that what you did here won't work (can you write a test?). Have you verified that it still works as expected?

My expectation is that it won't work, since you are trying to access files held in a zip (jar file, which is just a zip). 


<!-- Sent from Reviewable.io -->
